scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -2218,12 +2298,114 @@ static bool layered_monitor(void)
 	return true;
 }
 
+u64 antistall_set(u64 dsq_id, u64 now_sec) 
+{
+	struct task_struct *p;
+	struct task_ctx *tctx;
+	s32 cpu;
+	u64 *antistall_dsq, *delay, cur_delay, empty, runnable_at_sec;
+	bool first_pass;
+	u32 zero;
+	
+	zero = 0;
+	empty = -1;
+
+	if (!dsq_id || !now_sec)
+		return 0;
+	bpf_rcu_read_lock();	
+	bpf_for_each(scx_dsq, p, dsq_id, 0) {
+		
+		if (!(tctx = lookup_task_ctx(p)))
+			goto unlock;
+		
+		runnable_at_sec = p->scx.runnable_at / CONFIG_HZ;
+		cur_delay = now_sec - runnable_at_sec;
+		if (cur_delay <= antistall_sec || cur_delay > 100)
+			// check head task in dsq
+			goto unlock;
+		
+		first_pass = true;
+	look_for_cpu:
+		bpf_for(cpu, 0, nr_possible_cpus) {
+			if (!tctx->layered_cpumask)
+				goto unlock;
+			
+			if (!bpf_cpumask_test_cpu(cpu,   cast_mask(tctx->layered_cpumask)))
+				continue;
+			
+			antistall_dsq = bpf_map_lookup_percpu_elem(&antistall_cpu_dsq, &zero, cpu);
+			delay = bpf_map_lookup_percpu_elem(&antistall_cpu_max_delay, &zero, cpu);
+			
+			if (!delay || !antistall_dsq){
+				scx_bpf_error("cant happen");
+				goto unlock;
+			}
+
+			if (*delay == empty || *antistall_dsq == empty) {
+				trace("antistall set DSQ[%llu] SELECTED_CPU[%llu] DELAY[%llu]", dsq_id, cpu, cur_delay);
+				*delay = cur_delay;
+				*antistall_dsq = dsq_id;
+				goto unlock;
+			}
+
+			if (first_pass)
+				continue;
+				
+			if (*delay < cur_delay){
```

Comment:
Ah, okay, you're picking the more delayed one. Still, I think it'd be less confusing to use `SCX_DSQ_INVALID` as the empty marker and use `antistall_cpu_max_delay` only in this function.